File: 11-1.py
#!/usr/bin/env python

import numpy

class Map:
    def __init__(self, input):
        self.size = len(input)
        self.map = numpy.full((self.size, self.size), 0, dtype=numpy.uint)

        for row in range(self.size):
            for col in range(self.size):
                self.map[row, col] = input[row][col]

        self.flashed = numpy.full((self.size, self.size), 0, dtype=numpy.uint)
        self.num_flashes = 0

    def increase(self, row, col):
        if 0 <= row < self.size and \
           0 <= col < self.size:
            self.map[row, col] += 1

    def energize(self):
        for row in range(self.size):
            for col in range(self.size):
                self.increase(row, col)

    def flash(self):
        flashed = False
        for row in range(self.size):
            for col in range(self.size):
                if self.map[row, col] <= 9 or self.flashed[row, col] == 1:
                    continue

                self.num_flashes += 1
                flashed = True
                self.flashed[row, col] = 1
                self.increase(row - 1, col - 1)
                self.increase(row - 1, col)
                self.increase(row - 1, col + 1)
                self.increase(row, col - 1)
                self.increase(row, col + 1)
                self.increase(row + 1, col - 1)
                self.increase(row + 1, col)
                self.increase(row + 1, col + 1)

        return flashed

    def reset(self):
        for row in range(self.size):
            for col in range(self.size):
                if self.flashed[row, col] == 1:
                    self.map[row, col] = 0

        self.flashed.fill(0)

    def step(self):
        self.energize()
        while True:
            if not self.flash():
                break
        self.reset()

    def __str__(self):
        return f"<{self.__class__.__name__}:\nsize: {self.size}\n{self.map}\n>"


input = [list(line.strip()) for line in open('11.input').readlines()]
map = Map(input)

for i in range(100):
    map.step()

print(map.num_flashes)

File: 11-1.rb
#!/usr/bin/env ruby

require 'numo/narray'

class Map
  attr_accessor :num_flashes

  def initialize(input)
    @size = 10
    @map = Numo::UInt8.zeros(@size, @size)
    input.each_index do |row|
      input[row].each_index do |col|
        @map[row, col] = input[row][col]
      end
    end
    @flashed = Numo::UInt8.zeros(@size, @size)
    @num_flashes = 0
  end

  def increase(row, col)
    return unless row.between?(0, @size - 1)
    return unless col.between?(0, @size - 1)

    @map[row, col] += 1
  end

  def energize
    @size.times do |row|
      @size.times do |col|
        increase(row, col)
      end
    end
  end

  def flash
    flashed = false
    @size.times do |row|
      @size.times do |col|
        next unless @map[row, col] > 9 and
          @flashed[row, col].zero?

        @num_flashes += 1
        flashed = true
        @flashed[row, col] = 1
        increase(row - 1, col - 1)
        increase(row - 1, col)
        increase(row - 1, col + 1)
        increase(row, col - 1)
        increase(row, col + 1)
        increase(row + 1, col - 1)
        increase(row + 1, col)
        increase(row + 1, col + 1)
      end
    end
    flashed
  end

  def reset
    @size.times do |row|
      @size.times do |col|
        @map[row, col] = 0 if @flashed[row, col] == 1
      end
    end
    @flashed.fill 0
  end

  def step
    energize
    loop do
      break unless flash
    end
    reset
  end

  def to_s
    s = "<#{self.class}:\n"
    @size.times do |row|
      @size.times do |col|
        s += "#{@map[row, col]} "
      end
      s += "\n"
    end
    s += '>'
    s
  end

  def inspect
    to_s
  end
end

input = File.read('11.input').lines.map(&:strip).map { |row| (row.split '').map &:to_i }

map = Map.new(input)

100.times { |_| map.step }

print map.num_flashes, "\n"
